solc/CommandLineInterface.cpp

```diff
@@ -760,8 +763,78 @@ bool CommandLineInterface::processInput()
 			m_compiler->useMetadataLiteralSources(true);
 		if (m_args.count(g_argInputFile))
 			m_compiler->setRemappings(m_args[g_argInputFile].as<vector<string>>());
-		for (auto const& sourceCode: m_sourceCodes)
-			m_compiler->addSource(sourceCode.first, sourceCode.second);
+
+		if (m_args.count(g_argImportAst))
+		{
+			//read file contents, which either are
+			// a) a json-file with multiple sources
+			// b) a json-file with one source only
+			// c) a .sol file that corresponds to a given json
+			Json::Reader reader;
+			map<string, Json::Value const*> sourceJsons; // will be given to the compilerimportfunction
+			map<string, string> tmp_sources; //used to generate the onchainmetadata-hash
+			map<string, string> supplementarySourceCodes;
+			for (auto& srcPair: m_sourceCodes) // aka <string, string>
+			{
+				Json::Value* ast = new Json::Value(); //use shared-Pointer here?
+				// try parsing as json
+				if (reader.parse(srcPair.second, *ast, false))
+				{
+//					BOOST_THROW_EXCEPTION(InvalidAstError() << errinfo_comment("could not be parsed to JSON")); //TODO what to throw
+					// case a)
+					if ((*ast).isMember("sourceList") && (*ast).isMember("sources"))
+					{
+					for (auto& src: (*ast)["sourceList"]) //check here for ast-correctness as well
+						{
+							solAssert((*ast)["sources"][src.asCString()]["AST"]["nodeType"] == "SourceUnit", "the top-level node of the AST needs to be a 'SourceUnit'");
+							sourceJsons[src.asCString()] = &((*ast)["sources"][src.asCString()]["AST"]);
+							tmp_sources[src.asCString()] = "noSourceCodeAvailable";
+						}
+					}
+					// case b) WATCHOUT: the name right now is .json, yet it should be saved under .sol
+					else
+					{
+						solAssert((*ast)["nodeType"] == "SourceUnit", "invalid AST: the top-level node should be a SourceUnit");
+						string s = srcPair.first;
+						size_t f = s.find(".json");
+						s.replace(f, std::string(".json").length(), ".sol");
+						sourceJsons[s] = ast;
+						tmp_sources[s] = "noSourceCodeAvailable";
+						//TODO now the xxx.json is the absolute path name
+					}
+				}
+				else // c) supplementary .sol file
+				{
+					// add .sol to the import-input only if there already is a json-ast with a correspoding name
+					if (sourceJsons.count(srcPair.first)) //TODO-> should not work for case b) yet in which the name is name.json instead of name.sol
+						supplementarySourceCodes[srcPair.first] = srcPair.second;
+				}
+			}
+			//replace the internal map so that every source has its own entry
+			m_sourceCodes = tmp_sources;
+			//feed AST to compiler
+			m_compiler->reset(false);
+			bool import = m_compiler->importASTs(sourceJsons);
+			if (supplementarySourceCodes.size() > 0)
+				m_compiler->saveImportedSourceCodes(supplementarySourceCodes);
```

Comment:
when we wanted to add the .sol file in the metadata, we would have had to...Now that we decided to use the printed JSON as sourcefile, not anymore.